### PR TITLE
New text "background" setting for marker symbol backgrounds

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows for rendering a marker symbol in the background of labels (complementing the existing shapes/SVG choices). This allows use of QGIS marker symbology as a background symbol behind labels (e.g. for highway shield labels)

Plugs another symbology/labeling feature gap between ArcGIS and QGIS!

image

### CODE REVIEW

**Comment by nirvn** [comment]
You are on fire 🔥

File: src/core/qgstextrenderer.cpp [hunk 1]
```diff
@@ -861,6 +869,12 @@ void QgsTextBackgroundSettings::readXml( const QDomElement &elem, const QgsReadW
     setPaintEffect( QgsApplication::paintEffectRegistry()->createEffect( effectElem ) );
   else
     setPaintEffect( nullptr );
+
+  const QDomElement symbolElem = backgroundElem.firstChildElement( QStringLiteral( "symbol" ) );
```
Comment (nirvn) [inline src/core/qgstextrenderer.cpp]: Is "symbol" overly generic a string? No string feeling, just asking :)

File: src/core/qgstextrenderer.cpp [hunk 2]
```diff
@@ -892,8 +906,12 @@ QDomElement QgsTextBackgroundSettings::writeXml( QDomDocument &doc, const QgsRea
   backgroundElem.setAttribute( QStringLiteral( "shapeJoinStyle" ), static_cast< unsigned int >( d->joinStyle ) );
   backgroundElem.setAttribute( QStringLiteral( "shapeOpacity" ), d->opacity );
   backgroundElem.setAttribute( QStringLiteral( "shapeBlendMode" ), QgsPainting::getBlendModeEnum( d->blendMode ) );
-  if ( d->paintEffect && !QgsPaintEffectRegistry::isDefaultStack( d->paintEffect ) )
+  if ( d->paintEffect && !QgsPaintEffectRegistry::isDefaultStack( d->paintEffect.get() ) )
     d->paintEffect->saveProperties( doc, backgroundElem );
+
+  if ( d->markerSymbol )
+    backgroundElem.appendChild( QgsSymbolLayerUtils::saveSymbol( QStringLiteral( "marker" ), d->markerSymbol.get(), doc, context ) );
```
Comment (nirvn) [inline src/core/qgstextrenderer.cpp]: QStringLiteral( "symbol" ) ? I'd favor "marker" over "symbol", maybe change the readXml string?

File: src/core/qgstextrenderer.cpp [hunk 3]
```diff
@@ -2198,264 +2216,291 @@ void QgsTextRenderer::drawBackground( QgsRenderContext &context, QgsTextRenderer
...
-      sizeOut = std::max( component.size.width(), component.size.height() );
-      double bufferSize = context.convertToPainterUnits( background.size().width(), background.sizeUnit(), background.sizeMapUnitScale() );
+      if ( background.type() == QgsTextBackgroundSettings::ShapeMarkerSymbol && !background.markerSymbol() )
+        return;
 
-      // add buffer
-      sizeOut += bufferSize * 2;
-    }
+      double sizeOut = 0.0;
+      // only one size used for SVG sizing/scaling (no use of shapeSize.y() or Y field in gui)
```
Comment (nirvn) [inline src/core/qgstextrenderer.cpp]: for SVG and Marker sizing/scaling, right?

**Review by nirvn — COMMENTED** [review]
Looking good.

File: src/core/qgstextrenderer.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgstextrenderer.cpp]: It's baked into how symbols are saved -- the alternative would be to embed another depth of element, which I'd rather avoid

**Pull request merged**